### PR TITLE
[BC] Generate trajectories

### PR DESCRIPTION
**Pull request by tvmarino**
Commit gen_trajectories() which is the function that loads the modules corpus, creates the worker manager to be used with ModuleWorker, collects the results and writes the results to file.

### CODE REVIEW

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 1]
```diff
@@ -809,3 +827,162 @@ def select_best_exploration(
...
+      cps.load_module_spec(corpus_element) for corpus_element in corpus_elements
+  ]
+
+  modules_processed = 0
+  time_compiler_calls = 0
+  total_successful_examples = 0
+  total_work = len(corpus_elements)
+  total_failed_examples = 0
+  total_write_files = num_output_files
+  total_profiles_max: List[Optional[Dict[str, Union[str, float, int]]]] = []
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:878]: there is a lot of nesting. A trick to help comprehension is to alias the type somewhere (as a module-level def), e.g.: (I'm making the names up)

```
ExperimentValueType = Union[str, float, int]
ExperimentResultType = Dict[str, ExperimentValueType]
```

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 2]
```diff
@@ -809,3 +827,162 @@ def select_best_exploration(
...
+  work = [
+      cps.load_module_spec(corpus_element) for corpus_element in corpus_elements
+  ]
+
+  modules_processed = 0
+  time_compiler_calls = 0
+  total_successful_examples = 0
+  total_work = len(corpus_elements)
+  total_failed_examples = 0
+  total_write_files = num_output_files
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:877]: why not leave this as num_output_files? you're never mutating it

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 3]
```diff
@@ -809,3 +827,162 @@ def select_best_exploration(
...
+        jobs=work,
+        worker_pool=lwm,
+    )
+    not_done = result_futures
+    succeeded_idx = 0
+    succeeded: List[Optional[Tuple[Tuple[int, Dict[str, Union[str, float, int]],
+                                         Dict[str, Union[str, float, int]]],
+                                   tf.train.SequenceExample]]] = []
+
+    for written_files in range(total_write_files):
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:900]: s/written_files/written_file_index?

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 4]
```diff
@@ -809,3 +827,162 @@ def select_best_exploration(
...
+            logging.info('Module failed with: %s', f.exception())
+          total_successful_examples += len(succeeded)
+          total_failed_examples += len(done) - len(succeeded)
+          modules_processed += len(done)
+          while written_per_file < size_per_file:
+            logging.log_every_n_seconds(
+                logging.INFO,
+                ('%d success, %d failed out of %d, modules processed'
+                 ' %d\n timing compiler: %f'),
+                10,
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:929]: what's 10?

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 5]
```diff
@@ -809,3 +827,162 @@ def select_best_exploration(
...
+                total_successful_examples,
+                total_failed_examples,
+                total_work,
+                modules_processed,
+                time_compiler_calls,
+            )
+            if len(succeeded) == 0:
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:936]: doesn't `if not succeeded` work?

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 6]
```diff
@@ -809,3 +827,162 @@ def select_best_exploration(
...
+            if tfrecord_writer:
+              tfrecord_writer.write(records)
+            written_per_file += 1
+            succeeded_idx += 1
+          if written_per_file >= size_per_file:
+            break
+
+  max_profiles_path = ''
+  pol_profiles_path = ''
+  if profiling_file_path:
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:951]: so if `profiling_file_path` isn't given, then what happens with the `open` below?

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 2 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:877]: That's true. I will fix this.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 4 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:929]: From the log_every_n_seconds docstring it looks like it's the time between each logging.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 6 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:951]: The context is set to contextlib.nullcontext(), so I think nothing happens.

Comment (mtrofin) [reply compiler_opt/rl/generate_bc_trajectories.py:951]: would it be cleaner to check here and do the `open` stuff only when `profiling_file_path` is set?

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 7]
```diff
@@ -809,3 +827,162 @@ def select_best_exploration(
...
+    data_path: str = gin.REQUIRED,
+    delete_flags: Tuple[str, ...] = gin.REQUIRED,
+    output_file_name: str = gin.REQUIRED,
+    output_path: str = gin.REQUIRED,
+    obs_action_spec: Optional[Tuple[time_step.TimeStep,
+                                    tensor_spec.BoundedTensorSpec]] = None,
+    num_workers: Optional[int] = None,
+    num_output_files: int = 1,
+    profiling_file_path: Optional[str] = None,
+    worker_wait: int = 10,
```
[the remark this answers is not in the document]
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:843]: Done.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 6 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:951]: Yes, I am addressing this.

**Review by mtrofin — APPROVED** [review]

**Pull request merged**